### PR TITLE
require JSX files that can be covered by istanbul.

### PR DESCRIPTION
**Pull request by stanleyhlng**
@akshayp 

I am using `jsx-test` to test my react components and I would like to instrument the "transpiled" `.jsx` files and show up in the coverage report.  This PR:

- update the `jsxTranspile` method to support `.jsx` file instrumentation using [istabual](1).
- update npm run-script targets using [jenkin-mocha](2).

**1. The Walkthrough**
jsx-test

**2. Report for JSX file**
screen shot 2015-06-20 at 12 51 32 am

**3. Test with Coverage Result**
```bash
➜  jsx-test git:(instrumented-jsx-files) npm run cover

> jsx-test@0.6.3 cover /Users/stanleyn/Projects/public/git/yahoo/jsx-test
> jenkins-mocha example/**/*.test.js test/**/*.test.js



  CheckboxWithLabel
    ✓ changes the text after click
    ✓ triggers native events
    #render
      ✓ renders with a custom label OFF
      ✓ does not render a the unchecked label
      ✓ renders with a custom label ON
      ✓ renders with a checkbox field

  UnnamedComponent
    #render
      ✓ renders with <div>UNNAMED_COMPONENT</div>

  #assertRender
    ✓ renders the exact HTML
    ✓ renders the attomic classes HTML
    ✓ should handle data-react-* truncation correctly
    ✓ includes some expecific content on the rendered HTML
    ✓ matches a regex on the rendered HTML

  #renderComponent
    ✓ renders props & children on the component

  #stubComponent
    ✓ renders props on the component
    ✓ renders children content the component
    ✓ adds all props to data props

  #withContext
    ✓ gives child context
    ✓ passes props through context wrapper to child
    ✓ allows calling methods from underlying component
    ✓ creates a readable displayName using Component.displayName
    ✓ creates a readable displayName even if Component did not have one


  21 passing (82ms)

=============================================================================
Writing coverage object [/Users/stanleyn/Projects/public/git/yahoo/jsx-test/artifacts/coverage/coverage.json]
Writing coverage reports at [/Users/stanleyn/Projects/public/git/yahoo/jsx-test/artifacts/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 96.81% ( 91/94 )
Branches     : 83.33% ( 15/18 )
Functions    : 93.33% ( 28/30 )
Lines        : 96.81% ( 91/94 )
================================================================================
```

**4. Test without Coverage Result**
```bash
➜  jsx-test git:(instrumented-jsx-files) npm run  test

> jsx-test@0.6.3 test /Users/stanleyn/Projects/public/git/yahoo/jsx-test
> jenkins-mocha --no-coverage example/**/*.test.js test/**/*.test.js



  CheckboxWithLabel
    ✓ changes the text after click
    ✓ triggers native events
    #render
      ✓ renders with a custom label OFF
      ✓ does not render a the unchecked label
      ✓ renders with a custom label ON
      ✓ renders with a checkbox field

  UnnamedComponent
    #render
      ✓ renders with <div>UNNAMED_COMPONENT</div>

  #assertRender
    ✓ renders the exact HTML
    ✓ renders the attomic classes HTML
    ✓ should handle data-react-* truncation correctly
    ✓ includes some expecific content on the rendered HTML
    ✓ matches a regex on the rendered HTML

  #renderComponent
    ✓ renders props & children on the component

  #stubComponent
    ✓ renders props on the component
    ✓ renders children content the component
    ✓ adds all props to data props

  #withContext
    ✓ gives child context
    ✓ passes props through context wrapper to child
    ✓ allows calling methods from underlying component
    ✓ creates a readable displayName using Component.displayName
    ✓ creates a readable displayName even if Component did not have one


  21 passing (78ms)
```

[1]: https://github.com/gotwarlost/istanbul
[2]: https://github.com/stjohnjohnson/jenkins-mocha

### CODE REVIEW

File: lib/helper.js [hunk 1]
```diff
@@ -143,11 +143,20 @@ function withContext(Component, context) {
 function jsxTranspile() {
     var fs = require('fs');
     var babel = require('babel');
+    var istanbul = require('istanbul');
```
Comment (3den) [inline lib/helper.js]: @stanleyhlng can you make this optional?

**Comment by 3den** [comment]
It would be good to make code coverage optional, we can eather create a separated method for that or just pass flag to jsxTranspile.

**Comment by stanleyhlng** [comment]
@3den 

I have made the code coverage optional with a flag.

**Comment by 3den** [comment]
@stanleyhlng awesome! thanks :+1:

**Comment by 3den** [comment]
@stanleyhlng I am getting this error when i try to `npm install`:

```
➜  jsx-test git:(master) ✗ npm i       
 
> contextify@0.1.14 install /Users/eden/Code/javascript/jsx-test/node_modules/jsdom/node_modules/contextify
> node-gyp rebuild

gyp WARN install got an error, rolling back install
gyp ERR! configure error 
gyp ERR! stack Error: 404 response dowloading http://nodejs.org/dist/v2.0.2/node-v2.0.2.tar.gz
gyp ERR! stack     at Request.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/install.js:251:14)
gyp ERR! stack     at emitOne (events.js:82:20)
gyp ERR! stack     at Request.emit (events.js:169:7)
gyp ERR! stack     at Request.onRequestResponse (/usr/local/lib/node_modules/npm/node_modules/request/request.js:1142:10)
gyp ERR! stack     at emitOne (events.js:77:13)
gyp ERR! stack     at ClientRequest.emit (events.js:169:7)
gyp ERR! stack     at HTTPParser.parserOnIncomingClient (_http_client.js:414:21)
gyp ERR! stack     at HTTPParser.parserOnHeadersComplete (_http_common.js:88:23)
gyp ERR! stack     at Socket.socketOnData (_http_client.js:304:20)
gyp ERR! stack     at emitOne (events.js:77:13)
gyp ERR! System Darwin 14.3.0
gyp ERR! command "/usr/local/Cellar/iojs/2.0.2/bin/iojs" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/eden/Code/javascript/jsx-test/node_modules/jsdom/node_modules/contextify
gyp ERR! node -v v2.0.2
gyp ERR! node-gyp -v v2.0.1
gyp ERR! not ok 
npm ERR! Darwin 14.3.0
npm ERR! argv "/usr/local/Cellar/iojs/2.0.2/bin/iojs" "/usr/local/bin/npm" "i"
npm ERR! node v2.0.2
npm ERR! npm  v2.11.2
npm ERR! code ELIFECYCLE

npm ERR! contextify@0.1.14 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the contextify@0.1.14 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the contextify package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls contextify
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/eden/Code/javascript/jsx-test/npm-debug.log
```

**Comment by stanleyhlng** [comment]
@3den 

I am on node v0.10

```bash
➜  jsx-test git:(master) npm i

> contextify@0.1.14 install /Users/stanleyn/Temp/jsx-test/node_modules/jsdom/node_modules/contextify
> node-gyp rebuild

  CXX(target) Release/obj.target/contextify/src/contextify.o
  SOLINK_MODULE(target) Release/contextify.node
ld: warning: directory not found for option '-L/opt/local/lib'
  SOLINK_MODULE(target) Release/contextify.node: Finished

> fsevents@0.3.6 install /Users/stanleyn/Temp/jsx-test/node_modules/babel/node_modules/chokidar/node_modules/fsevents
> node-gyp rebuild

  SOLINK_MODULE(target) Release/.node
ld: warning: directory not found for option '-L/opt/local/lib'
  SOLINK_MODULE(target) Release/.node: Finished
  CXX(target) Release/obj.target/fse/fsevents.o
  SOLINK_MODULE(target) Release/fse.node
ld: warning: directory not found for option '-L/opt/local/lib'
  SOLINK_MODULE(target) Release/fse.node: Finished
jenkins-mocha@2.2.0 node_modules/jenkins-mocha
├── spec-xunit-file@0.0.1-3
├── shelljs@0.3.0
├── npm-which@2.0.0 (npm-path@1.0.2, commander@2.8.1, which@1.1.1)
└── mocha@2.2.5 (escape-string-regexp@1.0.2, supports-color@1.2.1, growl@1.8.1, commander@2.3.0, diff@1.4.0, debug@2.0.0, mkdirp@0.5.0, glob@3.2.3, jade@0.26.3)

jsdom@3.1.2 node_modules/jsdom
├── xmlhttprequest@1.7.0
├── xml-name-validator@1.0.0
├── browser-request@0.3.3
├── cssom@0.3.0
├── nwmatcher@1.3.4
├── parse5@1.4.2
├── contextify@0.1.14 (bindings@1.2.1, nan@1.8.4)
├── htmlparser2@3.8.3 (domelementtype@1.3.0, entities@1.0.0, domhandler@2.3.0, readable-stream@1.1.13, domutils@1.5.1)
├── acorn-globals@1.0.4 (acorn@1.2.2)
├── request@2.58.0 (caseless@0.10.0, aws-sign2@0.5.0, forever-agent@0.6.1, stringstream@0.0.4, tunnel-agent@0.4.0, oauth-sign@0.8.0, isstream@0.1.2, extend@2.0.1, json-stringify-safe@5.0.1, node-uuid@1.4.3, combined-stream@1.0.5, qs@3.1.0, mime-types@2.0.14, http-signature@0.11.0, bl@0.9.4, tough-cookie@2.0.0, form-data@1.0.0-rc1, hawk@2.3.1, har-validator@1.7.1)
├── escodegen@1.6.1 (esutils@1.1.6, estraverse@1.9.3, optionator@0.5.0, source-map@0.1.43, esprima@1.2.5)
├── cssstyle@0.2.28
└── acorn@0.11.0

istanbul@0.3.15 node_modules/istanbul
├── supports-color@1.3.1
├── which@1.0.9
├── abbrev@1.0.7
├── nopt@3.0.2
├── wordwrap@0.0.3
├── once@1.3.2 (wrappy@1.0.1)
├── async@0.9.2
├── mkdirp@0.5.1 (minimist@0.0.8)
├── esprima@2.1.0
├── resolve@1.1.6
├── fileset@0.1.6 (minimatch@0.4.0, glob@5.0.10)
├── handlebars@3.0.0 (optimist@0.6.1, source-map@0.1.43, uglify-js@2.3.6)
├── escodegen@1.6.1 (esutils@1.1.6, estraverse@1.9.3, optionator@0.5.0, source-map@0.1.43, esprima@1.2.5)
└── js-yaml@3.3.1 (esprima@2.2.0, argparse@1.0.2)

react@0.13.3 node_modules/react
└── envify@3.4.0 (through@2.3.7, jstransform@10.1.0)

babel@5.6.1 node_modules/babel
├── slash@1.0.0
├── path-is-absolute@1.0.0
├── fs-readdir-recursive@0.1.2
├── convert-source-map@1.1.1
├── commander@2.8.1 (graceful-readlink@1.0.1)
├── output-file-sync@1.1.1 (xtend@4.0.0, mkdirp@0.5.1)
├── glob@5.0.10 (inherits@2.0.1, once@1.3.2, inflight@1.0.4, minimatch@2.0.8)
├── source-map@0.4.2 (amdefine@0.1.1)
├── lodash@3.9.3
├── chokidar@1.0.3 (arrify@1.0.0, is-glob@1.1.3, glob-parent@1.2.0, async-each@0.1.6, is-binary-path@1.0.1, readdirp@1.3.0, anymatch@1.3.0, fsevents@0.3.6)
└── babel-core@5.6.1 (babel-plugin-remove-console@1.0.1, babel-plugin-remove-debugger@1.0.1, babel-plugin-jscript@1.0.1, babel-plugin-eval@1.0.1, babel-plugin-inline-environment-variables@1.0.1, babel-plugin-property-literals@1.0.1, babel-plugin-react-constant-elements@1.0.3, babel-plugin-member-expression-literals@1.0.1, babel-plugin-undefined-to-void@1.1.6, shebang-regex@1.0.0, trim-right@1.0.0, to-fast-properties@1.0.1, babel-plugin-react-display-name@1.0.3, babel-plugin-constant-folding@1.0.1, babel-plugin-undeclared-variables-check@1.0.2, babel-plugin-proto-to-assign@1.0.3, babel-plugin-dead-code-elimination@1.0.1, strip-json-comments@1.0.2, leven@1.0.2, babel-plugin-runtime@1.0.7, private@0.1.6, globals@6.4.1, estraverse@4.1.0, esutils@2.0.2, home-or-tmp@1.0.0, line-numbers@0.2.0, js-tokens@1.0.1, debug@2.2.0, chalk@1.0.0, detect-indent@3.0.1, repeating@1.1.3, ast-types@0.7.6, is-integer@1.0.4, minimatch@2.0.8, resolve@1.1.6, source-map-support@0.2.10, acorn-jsx@1.0.3, regexpu@1.2.0, regenerator@0.8.30, core-js@0.9.18)

➜  jsx-test git:(master) node --version
v0.10.38
```

**Comment by stanleyhlng** [comment]
by the way, you can update your .travis.yml file to support the build with multiple node_js version.

e.g.
```javascript
matrix:
  allow_failures:
  - node_js: '0.13'
node_js:
- iojs
- '0.13'
- '0.12'
- '0.10'
```

File: lib/helper.js [hunk 2]
```diff
@@ -139,16 +139,35 @@ function withContext(Component, context) {
...
+        var coverageVariable = Object.keys(global).filter(
+            function (key) {
+                return /^\$\$cov_/.test(key);
+            })[0];
+        var instrumenter = new istanbul.Instrumenter({
+                coverageVariable: coverageVariable
+            });
+        require.extensions['.jsx'] = function (module, filename) {
+            var content = fs.readFileSync(filename, 'utf8');
+            var compiled = babel.transform(content, {nonStandard: true});
```
Comment (akshayp) [inline lib/helper.js]: I think you also want `{retainLines: true}` here to get the report to be as accurate as it can be without source maps

File: lib/helper.js [hunk 3]
```diff
@@ -139,16 +139,35 @@ function withContext(Component, context) {
...
+            var instrumented = instrumenter.instrumentSync(compiled.code, filename);
```
Comment (akshayp) [inline lib/helper.js]: @stanleyhlng Files in the /tests/ and node_modules directory should not be instrumented so we should add a check here so they don't show up in the coverage report. You can do something trivial like `var ignoreFiles = /(tests|node_modules)/.test(filename);` or make it configurable in a dotfile from the project

File: lib/helper.js [hunk 4]
```diff
@@ -139,16 +139,35 @@ function withContext(Component, context) {
...
+function jsxTranspile(instrument) {
     var fs = require('fs');
     var babel = require('babel');
 
-    require.extensions['.jsx'] = function (module, filename) {
-        var content = fs.readFileSync(filename, 'utf8');
-        var compiled = babel.transform(content, {nonStandard: true});
-        return module._compile(compiled.code, filename);
-    };
+    if (instrument !== undefined && instrument && instrument === 'true') {
```
Comment (akshayp) [inline lib/helper.js]: if people use jenkins-mocha then you could also read --no-coverage from argv to determine this

File: lib/helper.js (hunk 2 above)
Comment (stanleyhlng) [inline lib/helper.js]: Let me try to add `{retainLins:true}`. will send another PR for that change.

File: lib/helper.js (hunk 3 above)
Comment (stanleyhlng) [inline lib/helper.js]: In my next PR, i will do inline for now. In the upcoming PR, we can read the dotfile, e.g. .jsx-test.yml.

ref implementation: https://github.com/nickmerwin/node-coveralls/blob/d5e1372bbd79a9fabbceec9624a220f1a8f75348/lib/getOptions.js#L72-L90

File: lib/helper.js (hunk 4 above)
Comment (stanleyhlng) [inline lib/helper.js]: @akshayp  i prefer `jenkins-mocha` setting an ENV variable for the coverage. e.g. JENKINS_MOCHA=[true | false].  Next, we can pass the ENV variable to the jsxTranspile().

```javacript
var jsx = require('../../index').jsxTranspile(process.env.JENKIN_MOCHA_COVERAGE);
```

Also, i dump the `process.argv` with jeknin-mocha, and I could not read --no-coverage from argv.  Here's the log.

```bash
[ '/usr/local/bin/node',
  '/Users/stanleyn/Projects/public/git/yahoo/jsx-test/node_modules/jenkins-mocha/node_modules/mocha/bin/_mocha',
  '--reporter',
  '/Users/stanleyn/Projects/public/git/yahoo/jsx-test/node_modules/jenkins-mocha/node_modules/spec-xunit-file',
  '--colors',
  'example/test/CheckboxWithLabel.test.js',
  'example/test/UnnamedComponent.test.js',
  'test/**/*.test.js' ]
[ '/usr/local/bin/node',
  '/Users/stanleyn/Projects/public/git/yahoo/jsx-test/node_modules/jenkins-mocha/node_modules/mocha/bin/_mocha',
  '--reporter',
  '/Users/stanleyn/Projects/public/git/yahoo/jsx-test/node_modules/jenkins-mocha/node_modules/spec-xunit-file',
  '--colors',
  'example/test/CheckboxWithLabel.test.js',
  'example/test/UnnamedComponent.test.js',
  'test/**/*.test.js' ]
```

**Comment by stanleyhlng** [comment]
@3den 

https://github.com/yahoo/jsx-test/pull/13#issuecomment-113870712

The error is come from `contextify` with `iojs` and you have `node v2.0.2` installed in your local dev env.
